Content.Server/Radio/EntitySystems/TelecomSystem.cs

```diff
@@ -1,5 +1,35 @@
+using Content.Server.Power.Components;
+using Content.Server.Radio.Components;
+using Content.Shared.Radio.Components;
+
 namespace Content.Server.Radio.EntitySystems;
 
 public sealed class TelecomSystem : EntitySystem
 {
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<RadioReceiveAttemptEvent>(OnRadioReceiveAttempt);
+    }
+
+    private void OnRadioReceiveAttempt(ref RadioReceiveAttemptEvent args)
+    {
+        if (args.Channel.longRange)
+            return;
+        var mapSource = Transform(args.RadioSource).MapID;
+        var mapReceiver = Transform(args.RadioReceiver).MapID;
+        if (mapSource == mapReceiver)
+        {
+            // if both source and receiver (for example) are handheld radios (but not headsets) then you don't server 
+            if (HasComp<RadioMicrophoneComponent>(args.RadioSource) && HasComp<RadioSpeakerComponent>(args.RadioReceiver))
+                return;
+            var servers = EntityQuery<TelecomServerComponent, EncryptionKeyHolderComponent, ApcPowerReceiverComponent, TransformComponent>();
+            foreach (var (_, keys, power, transform) in servers)
+            {
+                if (transform.MapID == mapSource && power.Powered && keys.Channels.Contains(args.Channel.ID))
+                    return;
+            }
+        }
+        args.Cancelled = true;
+    }
```

Comment:
I think you are gonna need to raise a cancellable `RadioSendAttemptEvent` at the start of  `SendRadioMessage` that does the query and then make receive just compare the map ids (though if you find a way to do it on send its better), that way the query gets ran every time you send a message and not on every single radio every time you send a message